protocols/kad/src/behaviour/test.rs

```diff
@@ -1317,3 +1317,96 @@ fn network_behaviour_inject_address_change() {
         kademlia.addresses_of_peer(&remote_peer_id),
     );
 }
+
+#[test]
+fn client_mode() {
+    // Create server peers.
+    let mut server_nodes = build_fully_connected_nodes_with_config(2, KademliaConfig::default());
+
+    // Create a client peer.
+    let mut cfg = KademliaConfig::default();
+    cfg.set_mode(Mode::Client);
+    let mut client = build_node_with_config(cfg);
+
+    // Fitler out peer Ids.
+    let mut peers: Vec<_> = server_nodes
+        .iter()
+        .map(|(_, swarm)| swarm.local_peer_id().clone())
+        .collect();
+    peers.push(client.1.local_peer_id().clone());
+
+    // Filter out MultiAddrs and swarms.
+    let mut addrs: Vec<_> = server_nodes.iter().map(|(addr, _)| addr.clone()).collect();
+    addrs.push(client.0.clone());
+
+    let mut swarms: Vec<_> = server_nodes.iter_mut().map(|(_, swarm)| swarm).collect();
+    swarms.push(&mut client.1);
+
+    let key = "123".to_string();
+    let key = Key::new(&key);
+    let record = Record::new(key.clone(), vec![4, 5, 6]);
+
+    // Put a record from a server peer.
+    swarms[0]
+        .behaviour_mut()
+        .put_record(record.clone(), Quorum::One)
+        .unwrap();
+
+    // Dial a server peer from the client peer.
+    swarms[2]
+        .behaviour_mut()
+        .add_address(&peers[1], addrs[1].clone());
+
+    // Try to get the same record from the client node.
+    swarms[2].behaviour_mut().get_record(&key, Quorum::One);
+
+    block_on(poll_fn(|ctx| {
+        for swarm in &mut swarms {
+            loop {
+                match swarm.poll_next_unpin(ctx) {
+                    Poll::Ready(Some(SwarmEvent::Behaviour(
+                        KademliaEvent::OutboundQueryCompleted { result, .. },
+                    ))) => match result {
+                        QueryResult::PutRecord(result) => match result {
+                            Ok(PutRecordOk { .. }) => {
+                                // Check if the server peer is not connected to the client peer.
+                                assert!(
+                                    swarm
+                                        .behaviour_mut()
+                                        .connected_peers
+                                        .iter()
+                                        .all(|p| *p != peers[2]),
+                                    "The server peer is connected to the client peer."
+                                );
+                            }
```

Comment:
How about calling `swarm[2].behaviour_mut().get_record(..)` once the first server is done with the `PutRecord` query. That would prevent any races between the PUT and the GET.

Reply:
I think, `swarm[2]` won't be available to us inside the `for` loop.

Reply:
Hi, I have a doubt with the present test we have:

> It is kind of hard to reliably check for absence of something, especially in an async setting.
> 1. The server stores a value `Kademlia::put_value`, waiting for a `KademliaEvent::OutboundQueryCompleted`.
> 2. The client connects to the server.
> 3. The client does a `Kademlia::get_record`, waiting for a `KademliaEvent::OutboundQueryCompleted`.
> 4. Check that the client has the server in its routing table and check that the server does not have the client in its routing table.

Are we sure that using the GetRecord operation will update the routing tables of the server peer? I'm trying to find the code, which does the "update"-ing. But, I'm unable to find it. Could you please point me towards the appropriate code?